src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -2458,6 +2458,14 @@ MulticopterPositionControl::calculate_velocity_setpoint(float dt)
 		vel_sp_slewrate(dt);
 	}
 
+	/* make sure velocity setpoint is constrained in all directions (xyz) */
+	float vel_norm_xy = sqrtf(_vel_sp(0) * _vel_sp(0) + _vel_sp(1) * _vel_sp(1));
+
+	if (vel_norm_xy > _vel_max_xy) {
+		_vel_sp(0) = _vel_sp(0) * _vel_max_xy / vel_norm_xy;
+		_vel_sp(1) = _vel_sp(1) * _vel_max_xy / vel_norm_xy;
+	}
+
 	_vel_sp_prev = _vel_sp;
```

Comment:
The constraint of the velocity setpoint should be at the end after any setpoint adjustement that could generally go worng.

Can you instead of moving the constraint just move the assignement line `_vel_sp_prev = _vel_sp;` to the end (after `_vel_sp(2) = math::constrain(_vel_sp(2), -_params.vel_max_up, _params.vel_max_down);`) and do your checks again?

I think it would make much more sense like that and you use only constrained setpoints for the slewrate.

Reply:
Actually it's the first thing I did. But (as I wrote in PR) it leads to smooth landing limiter problems: `_takeoff_vel_limit` never goes up because `_vel_sp_prev(2)` is now constrained by it (line 2469 in master) so in next cycle `vel_sp_slewrate()` will apply `max_acc_z` and `_vel_sp(2)` will be positive. It leads to indefinite decrease of `_takeoff_vel_limit` so taking off is impossible.

The alternative is to break `_vel_sp_prev = _vel_sp` assignment into separate assignments for x, y, z. So we will leave `_vel_sp_prev(2) = _vel_sp(2)` where it is now and move `_vel_sp_prev(0) = _vel_sp(0)` and `_vel_sp_prev(1) = _vel_sp(1)` to the end of function. However, in this case `_vel_sp_prev` is in mixed state between assignment which is quite awful.

The best fix would be to somehow change takeoff limiter behavior and move `_vel_sp_prev` to the end as you propose, but it's more time-consuming and will require more thorough testing. I'm going to open issue to discuss smooth takeoff limiter, but I believe this bug is really critical and should be fixed ASAP.

Reply:
Good explanation and reasoning. Are you sure the same issue does not happen in z direction?

Reply:
As quick smooth takeoff limiter fix it may be possible to replace
`_takeoff_vel_limit += -_vel_sp(2) * dt / _takeoff_ramp_time.get();` with something like
`_takeoff_vel_limit += _params.tko_speed * dt / _takeoff_ramp_time.get();`
I'm not sure it will not affect smooth takeoff logic, but as far as I saw in SITL `-_vel_sp(2)` always equals `_params.tko_speed` during takeoff.

Reply:
> Are you sure the same issue does not happen in z direction?
I thought about that too but tests in SITL had shown no overshoot (about 0.5 m overshoot for flight from 0 to -400 m at z-direction with 8 m/s). I think we should move `_vel_sp_prev` assignment to the end of function at some point later anyway so z-axis overshoot possibility will be eliminated.